src/preprocessing/passes/non_clausal_simp.cpp

```diff
@@ -448,10 +438,154 @@ PreprocessingPassResult NonClausalSimp::applyInternal(
         Rewriter::rewrite(Node(learnedBuilder)));
   }
 
-  propagator->setNeedsFinish(true);
   return PreprocessingPassResult::NO_CONFLICT;
 }  // namespace passes
 
+SatValue NonClausalSimp::solveByCryptominisat(
+    CVC4::prop::TseitinCnfStream* d_cnfStream,
+    AssertionPipeline* assertionsToPreprocess,
+    unsigned substs_index)
+{
+  for (size_t i = 0, size = assertionsToPreprocess->size(); i < size; ++i)
+  {
+    Trace("non-clausal-simplify") << "Assertion #" << i << " : "
+                                  << (*assertionsToPreprocess)[i] << std::endl;
+  }
+
+  Trace("skeleton-preprocessing") << "asserting to sat solver" << std::endl;
+  SatSolver* d_satSolver = d_cnfStream->getSatSolver();
+  { // for timing
+    TimerStat::CodeTimer timer(d_statistics.d_addingAssertions);
+    for (size_t i = 0, size = assertionsToPreprocess->size(); i < size; ++i)
+    {
+      Assert(Rewriter::rewrite((*assertionsToPreprocess)[i])
+             == (*assertionsToPreprocess)[i]);
+      // Don't reprocess substitutions
+      if (substs_index > 0 && i == substs_index)
+      {
+        continue;
+      }
+      Trace("skeleton-preprocessing")
+          << "asserting " << (*assertionsToPreprocess)[i] << std::endl;
+      d_cnfStream->convertAndAssert(
+          (*assertionsToPreprocess)[i], false, false, RULE_GIVEN);
+    }
+  } // end timing
+
+  return d_satSolver->solve();
+}
+
+std::tuple<bool, std::vector<Node>> NonClausalSimp::preprocessByCryptominisat(
+    AssertionPipeline* assertionsToPreprocess, unsigned substs_index)
+{
+  std::unique_ptr<SatSolver> d_satSolver(SatSolverFactory::createCryptoMinisat(
+      smtStatisticsRegistry(), "non_clausal_simp_solver"));
+  CVC4::prop::NullRegistrar d_registrar;
+  context::Context d_context;
+  CVC4::prop::TseitinCnfStream d_cnfStream(d_satSolver.get(),
+                                           &d_registrar,
+                                           &d_context,
+                                           true,
+                                           "toCNF-skeleton-preprocessing");
+  std::vector<Node> learned_literals;
+  if (solveByCryptominisat(&d_cnfStream, assertionsToPreprocess, substs_index)
+      == SAT_VALUE_FALSE)
+  {
+    // If in conflict, just return false
+    return std::make_tuple(false, learned_literals);
+  }
+
+  std::vector<SatLiteral> topLevelUnits = d_satSolver->getTopLevelUnits();
+  for (size_t i = 0; i < topLevelUnits.size(); i++)
+  {
+    SatLiteral lit = topLevelUnits[i];
+    if (d_cnfStream.getNodeCache().find(lit)
+        != d_cnfStream.getNodeCache().end())
```

Comment:
What does it mean if the literal is not in the node cache? Why is it safe to just skip those literals? Please add a comment.

Reply:
Right, that's what the if statement says :) However, I think that we need a deeper understanding of what is going on here. The node cache seems to be the literal to node map that CnfStream maintains. Now, the question is: are we expecting every literal that we get to be in that map? If yes, this should become an assertion. Are we expecting some literals to be missing from the map? If yes, we should document why and/or find out if that has a negative impact on our performance (since we are ignoring some learned literals in that case).

Reply:
I think we should expect every literal that we get to be in that map. getNodeCache() returns the "d_literalToNodeMap" of the cnfStream we created. 
We set the fullLitToNodeMap to true when creating the cnfStream. Therefore, when a new literal is created, whether it's a theory literal or not, it will be added to the d_literalToNodeMap and the SAT-solver.
https://github.com/CVC4/CVC4/blob/8b8e40985b83354e7e4a503d217374eb7fbc45f0/src/prop/cnf_stream.cpp#L197-L200

Reply:
I believe we should make it an assertion.

Reply:
Actually, one small twist. 
The SAT_solver always returns "True" and "Not False" (which correspond to satliteral 0 and ~1), in addition to the meaningful top level units. 
So I think I should start from index two when traversing the topLevelUnits

Reply:
Also should I have the assertion there? It seems a bit expensive. Doing that for each top level units is "n Log (n)"

Reply:
I would not start from index two, I would explicitly skip true/false lits. You can use the util function FindOrDie, to either get the element of a map or fail: https://github.com/CVC4/CVC4/blob/987df3df987768e2ce0c36d17469929f8e92fdec/src/base/map_util.h#L88

Reply:
I agree with @mpreiner, starting from index two is brittle because the order may be different for different SAT solvers.

> Also should I have the assertion that the topLevelUnit must be in the map there? It seems a bit expensive. Doing that for each top level units is "n Log (n)"

Assertions are only checked when compiling a debug version of CVC4 (or with assertions explicitly enabled), so we can and should use them liberally. Assertions document intent and they make CVC4 fail closer to an actual problem instead of later down the road. Also, note that `d_literalToNodeMap` is a hash map, so you'd expect a search to be on the order of O(1) (on average). Additionally, the check can be done without a second search, check out the implementation of `FindOrDie()`.

Could you also add the explanation for why we expect it to be in the cache as a comment in the source code?